pkg/client/client.go

```diff
@@ -0,0 +1,410 @@
+				c.impl.OnError(err)
+				continue
+			}
+
+			var wg sync.WaitGroup
+			wg.Add(2)
+			done := make(chan bool)
+
+			// expected state check-ins
+			go func() {
+				defer wg.Done()
+				for {
+					expected, err := checkinClient.Recv()
+					if err != nil {
+						if err != io.EOF {
+							c.impl.OnError(err)
+						}
+						close(done)
+						return
+					}
+
+					if c.expected == proto.StateExpected_STOPPING {
+						// in stopping state, do nothing with any other expected states
+						continue
+					}
+					if expected.State == proto.StateExpected_STOPPING {
+						// Elastic Agent is requesting us to stop.
+						c.expected = expected.State
+						c.impl.OnStop()
+						continue
+					}
+					if expected.ConfigStateIdx != c.cfgIdx {
+						// Elastic Agent is requesting us to update config.
+						c.cfgLock.Lock()
+						c.cfgIdx = expected.ConfigStateIdx
+						c.cfg = expected.Config
+						c.cfgLock.Unlock()
+						c.impl.OnConfig(expected.Config)
+						continue
+					}
+				}
+			}()
+
+			// observed state check-ins
+			go func() {
+				defer wg.Done()
+
+				var lastSent time.Time
+				var lastSentCfgIdx uint64
+				var lastSentStatus proto.StateObserved_Status
+				var lastSentMessage string
+				for {
+					select {
+					case <-done:
+						return
+					case <-time.After(500 * time.Millisecond):
+					}
+
+					c.cfgLock.RLock()
+					cfgIdx := c.cfgIdx
+					c.cfgLock.RUnlock()
+					c.obsLock.RLock()
+					observed := c.observed
+					observedMsg := c.observedMessage
+					c.obsLock.RUnlock()
+
+					sendMessage := func() {
+						err := checkinClient.Send(&proto.StateObserved{
+							Token:          c.token,
+							ConfigStateIdx: cfgIdx,
+							Status:         observed,
+							Message:        observedMsg,
+						})
+						if err != nil {
+							c.impl.OnError(err)
+							return
+						}
+						lastSent = time.Now()
+						lastSentCfgIdx = cfgIdx
+						lastSentStatus = observed
+						lastSentMessage = observedMsg
+					}
+
+					// On start keep trying to send the initial check-in.
+					if lastSent.IsZero() {
+						sendMessage()
+						continue
+					}
+
+					// Send new status when it has changed.
+					if lastSentCfgIdx != cfgIdx || lastSentStatus != observed || lastSentMessage != observedMsg {
+						sendMessage()
+						continue
+					}
+
+					// Send when more than 30 seconds has passed without any status change.
+					if time.Now().Sub(lastSent) >= c.minCheckTimeout {
+						sendMessage()
+						continue
+					}
+				}
+			}()
+
+			// wait for both send and recv go routines to stop before
+			// starting a new stream.
+			wg.Wait()
+		}
+	}()
+}
+
+// startActions starts the go rountines to send and receive actions
+func (c *Client) startActions() {
+	c.wg.Add(1)
+
+	// results our held outside of the retry loop, because on re-connect
+	// we still want to send the responses that either failed or haven't been
+	// sent back to the agent.
+	actionResults := make(chan *proto.ActionResponse, 100)
```

Comment:
why 100? maybe this should be sync and enable processing action after one is finished. sender then should handle timeouts in processing or send message after last one is confirmed, otherwise i think messages might be lost

Reply:
Just a high enough number to handle the actions. I think 100 is really high in general, I think it could be even 10, but didn't want it to cause any blocking issues.

I don't think we should serialize the actions at the client level, I think if an action can conflict with another action then the application that implements the actions should handle that part. Just because an action request is received doesn't mean it must handle it immediately, if another action is prevent that action.

The actionsResponse channel is outside of the loop so responses will not be lost during reconnects.